File: python/02_regret_analysis.py
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
from typing import Optional

def compute_regret(df: pd.DataFrame, oracle_policy: str = 'myopic-oracle') -> pd.DataFrame:
    """Compute empirical regret relative to the best observed reward per seed/scenario."""
    if 'cumulative_reward' not in df.columns:
        raise ValueError("DataFrame must contain 'cumulative_reward' column")

    if {'empirical_best_reward', 'empirical_regret'}.issubset(df.columns):
        regret_df = df.copy()
        regret_df['regret'] = regret_df['empirical_regret']
        regret_df['regret_baseline'] = regret_df['empirical_best_reward']
    elif {'oracle_reward', 'oracle_regret'}.issubset(df.columns):
        regret_df = df.copy()
        regret_df['regret'] = regret_df['oracle_regret']
        regret_df['regret_baseline'] = regret_df['oracle_reward']
    else:
        oracle_df = df.groupby(['seed', 'scenario'])['cumulative_reward'].max().reset_index()
        oracle_df.rename(columns={'cumulative_reward': 'regret_baseline'}, inplace=True)
        regret_df = pd.merge(df, oracle_df, on=['seed', 'scenario'], how='left')
        regret_df['regret'] = regret_df['regret_baseline'] - regret_df['cumulative_reward']
    
    # Avoid division by zero
    mask = regret_df['regret_baseline'] > 0
    regret_df.loc[mask, 'normalized_regret'] = regret_df.loc[mask, 'regret'] / regret_df.loc[mask, 'regret_baseline']
    
    return regret_df

def summarize_policy_rankings(regret_df: pd.DataFrame) -> pd.DataFrame:
    """Compute mean regret, average rank, and win rate per policy."""
    # Compute dense rank descending (1 is highest reward)
    regret_df['rank'] = regret_df.groupby(['seed', 'scenario'])['cumulative_reward'].rank(method='min', ascending=False)
    
    # Summarize
    summary = regret_df.groupby(['scenario', 'policy']).agg(
        mean_reward=('cumulative_reward', 'mean'),
        mean_regret=('regret', 'mean'),
        sd_regret=('regret', 'std'),
        mean_rank=('rank', 'mean'),
        win_rate=('rank', lambda x: (x == 1).mean())
    ).reset_index()
    
    return summary.sort_values(by=['scenario', 'mean_rank'])

def plot_regret_distribution(regret_df: pd.DataFrame) -> None:
    """Plot boxplots of regret per policy across scenarios."""
    plt.figure(figsize=(12, 8))
    
    # Order policies by median regret
    order = regret_df.groupby('policy')['regret'].median().sort_values().index
    
    sns.boxplot(
        data=regret_df, 
        x='policy', 
        y='regret', 
        hue='policy',
        order=order
    )
    
    # Create grid of subplots for each scenario (using catplot)
    g = sns.catplot(
        data=regret_df,
        x='policy',
        y='regret',
        hue='policy',
        col='scenario',
        kind='box',
        order=order,
        col_wrap=3,
        sharey=False,
        height=5,
        aspect=1.2
    )
    
    g.set_xticklabels(rotation=45, ha='right')
    g.set_axis_labels("Policy", "Empirical Regret (Best Observed - Policy Reward)")
    g.fig.suptitle("Regret Distribution by Policy and Scenario", y=1.05)
    
    return g.fig

File: R/02_regret_analysis.R
#' 02_regret_analysis.R
#' Functions to compute policy regret, ranking flips, and distributional behavior over sweeps.

library(dplyr)
library(tidyr)
library(ggplot2)

source("01_load_data.R")

#' Compute Empirical Regret
#'
#' For each seed and scenario, compute regret relative to the best observed reward.
#'
#' @param df Dataframe from load_parquet_results or load_compare_json
#' @param oracle_policy The name of the oracle policy.
#' @return Dataframe with computed regret columns.
compute_regret <- function(df, oracle_policy = "myopic-oracle") {

  if (all(c("empirical_best_reward", "empirical_regret") %in% names(df))) {
    regret_df <- df |>
      mutate(
        regret = empirical_regret,
        regret_baseline = empirical_best_reward,
        normalized_regret = ifelse(regret_baseline > 0, regret / regret_baseline, NA)
      )
  } else if (all(c("oracle_reward", "oracle_regret") %in% names(df))) {
    regret_df <- df |>
      mutate(
        regret = oracle_regret,
        regret_baseline = oracle_reward,
        normalized_regret = ifelse(regret_baseline > 0, regret / regret_baseline, NA)
      )
  } else {
    oracle_df <- df |>
      group_by(seed, scenario) |>
      summarize(regret_baseline = max(cumulative_reward, na.rm = TRUE), .groups = "drop")

    regret_df <- df |>
      left_join(oracle_df, by = c("seed", "scenario")) |>
      mutate(
        regret = regret_baseline - cumulative_reward,
        normalized_regret = ifelse(regret_baseline > 0, regret / regret_baseline, NA)
      )
  }

  return(regret_df)
}

#' Summarize Policy Rankings
#'
#' @param regret_df Dataframe with computed regret
#' @return Summary statistics showing mean regret and average ranking of policies
summarize_policy_rankings <- function(regret_df) {
  
  # Compute rankings per seed
  ranked_df <- regret_df |>
    group_by(seed, scenario) |>
    mutate(rank = rank(-cumulative_reward, ties.method = "min")) |>
    ungroup()
  
  # Aggregate stats
  summary_stats <- ranked_df |>
    group_by(scenario, policy) |>
    summarize(
      mean_reward = mean(cumulative_reward, na.rm = TRUE),
      mean_regret = mean(regret, na.rm = TRUE),
      sd_regret = sd(regret, na.rm = TRUE),
      mean_rank = mean(rank, na.rm = TRUE),
      win_rate = mean(rank == 1, na.rm = TRUE),
      .groups = "drop"
    ) |>
    arrange(scenario, mean_rank)
  
  return(summary_stats)
}

#' Plot Regret Distribution
#'
#' @param regret_df Dataframe with regret
#' @return ggplot object showing boxplots of regret per policy
plot_regret_distribution <- function(regret_df) {
  p <- ggplot(regret_df, aes(x = reorder(policy, regret, FUN = median), y = regret, fill = policy)) +
    geom_boxplot(outlier.alpha = 0.5) +
    facet_wrap(~ scenario, scales = "free_y") +
    theme_minimal() +
    labs(
      title = "Regret Distribution by Policy and Scenario",
      x = "Policy",
      y = "Empirical Regret (Best Observed - Policy Reward)"
    ) +
    theme(axis.text.x = element_text(angle = 45, hjust = 1), legend.position = "none")
  
  return(p)
}

# Example Usage:
# df <- load_compare_json("../results/my_compare_output.json")
# regret_df <- compute_regret(df)
# stats <- summarize_policy_rankings(regret_df)
# print(stats)
# p <- plot_regret_distribution(regret_df)
# ggsave("regret_plot.png", p, width=10, height=6)
